### PR TITLE
Expense CRUD operations

### PR DESCRIPTION
**Pull request by tstejl**
**expense**
* now displays bar chart of expense versus budget, aggregated be category
* be able to add/edit/delete expense
* *problem*
  * data table does not refresh upon CRUD operations, while the chart does
  * to be fixed together with progress problems

### CODE REVIEW

**Comment by magickaichen** [comment]
lgtm so far, will review this tomorrow

File: project_management/api/expenses/views.py [hunk 1]
```diff
@@ -72,17 +82,19 @@ def addExpense(request):
...
-        expense.project_id=request.data['project_id']
-        expense.workstream_id=request.data.get('workstream_id', None)
-        expense.activity_id=request.data.get('activity_id', None)
+        expense.description = request.data['description']
+        expense.category = request.data['category']
+        expense.amount = request.data['amount']
+        expense.date = request.data.get('date', None)
+        expense.approval_date = request.data.get('approval_date', None)
+        #expense.project_id = request.data['project_id']
+        #expense.workstream_id = request.data.get('workstream_id', None)
```
Comment (jouhyoung) [inline project_management/api/expenses/views.py]: no need to keep the comments here

File: project_management/api/expenses/views.py [hunk 2]
```diff
@@ -99,23 +111,94 @@ def deleteExpense(request):
...
+    """
+    given id of project
+    return dictionary of the following format:
+        category: {budget: ###, expense, ###}
+    """
+    budgets = Budget.objects.filter(project=id).values('category').\
+        annotate(amount=Sum('amount')).order_by('category')
+    budget = list(budgets)
+    for i in budget:
+        i['amount'] = round(float(i['amount']), 2)
```
Comment (jouhyoung) [inline project_management/api/expenses/views.py]: `for budget in budgets:`.
let's remove all variables `i`, `k`, `v` and give them names for readability.

File: project_management/api/expenses/views.py [hunk 3]
```diff
@@ -99,23 +111,94 @@ def deleteExpense(request):
...
+    expense = list(results)
+    for i in expense:
+        i['expense'] = round(float(i['expense']), 2)
+
+    dic = {}
+    for i in expense:
+        cat = i['category']
+        amo = i['expense']
+        if cat not in dic:
+            temp = {}
```
Comment (jouhyoung) [inline project_management/api/expenses/views.py]: what's temp?

File: project_management/api/expenses/views.py [hunk 4]
```diff
@@ -13,24 +13,32 @@ def expensesProject(request, id):
     """
     expenses = Expense.objects.filter(project=id)
     serializer = ExpenseSerializer(expenses, many=True)
+    for i in serializer.data:
```
Comment (jouhyoung) [inline project_management/api/expenses/views.py]: same here, lets not use i as variable

Comment (tstejl) [reply project_management/api/expenses/views.py]: @jouhyoung I've corrected all the `i`, `k`, `v`, and also some `temp` with comments
But for some simple logic is it okay to use one letter vars? for example `for k, v in dict.items()` `k` and `v` are somewhat standardized to mean keys and values

Comment (jouhyoung) [reply project_management/api/expenses/views.py]: k and v are standard, turning it to key and value doesn't help.
The situation with the naming is `dic` and `v` is, in fact, a dictionary as well. It'd be better to have a meaningful variable for `dic` and potentially better variable than `value`

Comment (tstejl) [reply project_management/api/expenses/views.py]: I'll think about `dic`. `v` has been commented when it has been assigned to var `cat_exp_bud`.
`dic` is simply used to remove duplicate categories from previous processes, and it contains category, expense and budget of the category.  I currently don't have a good idea on how to rename `dic` to resemble the contents inside, while keeping the name relatively short.

Comment (jouhyoung) [reply project_management/api/expenses/views.py]: regardless, I don't think this is a blocker.

**Pull request merged**